crates/primitives/src/result.rs

```diff
@@ -0,0 +1,134 @@
+    /// https://eips.ethereum.org/EIPS/eip-658
+    pub fn is_success(&self) -> bool {
+        matches!(self, Self::Success { .. })
+    }
+
+    /// Return logs, if execution is not successful, function will return empty vec.
+    pub fn logs(&self) -> Vec<Log> {
+        match self {
+            Self::Success { logs, .. } => logs.clone(),
+            _ => Vec::new(),
+        }
+    }
+
+    pub fn gas_used(&self) -> u64 {
+        let (Self::Success { gas_used, .. }
+        | Self::Revert { gas_used }
+        | Self::Halt { gas_used, .. }) = self;
+
+        *gas_used
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub enum Output {
+    #[cfg_attr(feature = "serde", serde(with = "crate::utilities::serde_hex_bytes"))]
+    Call(Bytes),
+    Create(
+        #[cfg_attr(feature = "serde", serde(with = "crate::utilities::serde_hex_bytes"))] Bytes,
+        Option<B160>,
+    ),
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub enum EVMError<DB> {
+    Transaction(InvalidTransaction),
+    /// REVM specific and related to environment.
+    PrevrandaoNotSet,
+    Database(DB),
+}
+
+impl<DB> From<InvalidTransaction> for EVMError<DB> {
+    fn from(invalid: InvalidTransaction) -> Self {
+        EVMError::Transaction(invalid)
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub enum InvalidTransaction {
+    GasMaxFeeGreaterThanPriorityFee,
+    GasPriceLessThenBasefee,
+    CallerGasLimitMoreThenBlock,
+    CallGasCostMoreThenGasLimit,
+    /// EIP-3607 Reject transactions from senders with deployed code
+    RejectCallerWithCode,
+    /// Transaction account does not have enough amount of ether to cover transferred value and gas_limit*gas_price.
+    LackOfFundForGasLimit,
+    /// Overflow payment in transaction.
+    OverflowPaymentInTransaction,
+    /// Nonce overflows in transaction,
+    NonceOverflowInTransaction,
+}
+
+/// When transaction return successfully without halts.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub enum Eval {
+    Stop,
+    Return,
+    SelfDestruct,
+}
+
+/// Indicates that the EVM has experienced an exceptional halt. This causes execution to
+/// immediately end with all gas being consumed.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub enum Halt {
+    OutOfGas,
+    OpcodeNotFound,
+    InvalidFEOpcode,
+    InvalidJump,
+    NotActivated,
+    StackUnderflow,
+    StackOverflow,
+    OutOfOffset,
+    CreateCollision,
+    OverflowPayment,
+    PrecompileError,
+    NonceOverflow,
+    /// Create init code size exceeds limit (runtime).
+    CreateContractSizeLimit,
```

Comment:
@rakita, I was looking at the reference implementation and am now wondering whether `CreateContractSizeLimit` is considered an exceptional halt:

https://github.com/ethereum/execution-specs/blob/95ef6bf957a4c66993168384cec395ef1a030026/src/ethereum/london/vm/exceptions.py#L115

Wouldn't we be able to catch that by validating the arguments passed to a create transaction at the very start?

Additionally, the `WriteInStaticContext` is considered an exceptional halt, but we treat is as an internal error. Why the difference? Can we guarantee recovery?

Reply:
`CreateContractSizeLimit` Not exactly, as data that is sent with a transaction is used as bytecode that is going to be executed, that bytecode can do anything and return of execution is newly created bytecode that is saved to db.

`WriteInStaticContext` Is not possible for a call of depth 1(as in the first call or create) it is only possible when call of depth 1 calls opcode STATIC_CALL for depth 2.